File: core/sdk/dotnet/Services/AgienceChatCompletionService.cs
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.ChatCompletion;

namespace Agience.Core.Services
{
    public class AgienceChatCompletionService : IChatCompletionService
    {
        public IReadOnlyDictionary<string, object?> Attributes => throw new NotImplementedException();

        private readonly KernelFunction _chatCompletionFunction;

        public AgienceChatCompletionService(KernelFunction chatCompletionFunction)
        {
            _chatCompletionFunction = chatCompletionFunction;
        }

        public async Task<IReadOnlyList<ChatMessageContent>> GetChatMessageContentsAsync(ChatHistory chatHistory, PromptExecutionSettings? executionSettings = null, Kernel? kernel = null, CancellationToken cancellationToken = default)
        {

            Data data = new Data();
            
            var args = new KernelArguments(executionSettings);

            args["chatHistory"] = chatHistory;
            args["executionSettings"] = executionSettings;
            args["agentId"] = kernel?.Data["agent_id"];
            //args["cancellationToken"] = cancellationToken;

            // TODO: We need to ensure that the Chat Completion function will return a list of ChatMessageContent. This will break otherwise.
            // TODO: Support other formats for input/output of the ChatCompletionFunction.

            return await _chatCompletionFunction.InvokeAsync<IReadOnlyList<ChatMessageContent>>(kernel, args, cancellationToken) ?? new List<ChatMessageContent>();
        }

        public IAsyncEnumerable<StreamingChatMessageContent> GetStreamingChatMessageContentsAsync(ChatHistory chatHistory, PromptExecutionSettings? executionSettings = null, Kernel? kernel = null, CancellationToken cancellationToken = default)
        {
            throw new NotImplementedException();
        }
    }
}

File: core/sdk/python/services/agience_chat_completion_service.py
import asyncio
import logging
from typing import Dict, List, Optional, AsyncIterable, Any

from semantic_kernel.kernel import Kernel
from semantic_kernel.functions import KernelFunction, FunctionResult
from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
from semantic_kernel.contents import ChatMessageContent, StreamingChatMessageContent
from semantic_kernel.contents.chat_history import ChatHistory
from semantic_kernel.connectors.ai.chat_completion_client_base import ChatCompletionClientBase
from semantic_kernel.functions.kernel_arguments import KernelArguments


class AgienceChatCompletionService(ChatCompletionClientBase):

    chat_completion_function: KernelFunction
    attributes: Dict[str, Any] = {}

    class Config:
        arbitrary_types_allowed = True

    async def get_chat_message_contents(
        self,
        kernel: Kernel,
        chat_history: ChatHistory,
        execution_settings: Optional[PromptExecutionSettings] = None,
        cancellation_token: Optional[asyncio.Event] = None,
        **kwargs,
    ) -> FunctionResult:
        if getattr(kernel, '_in_completion', False):
            return []

        try:
            # Set recursion guard
            kernel._in_completion = True

            args = KernelArguments(
                settings=execution_settings,
                chat_history=chat_history,
                agent_id=getattr(kernel, "agent_id", None) if kernel else None
            )

            result = await self.chat_completion_function.invoke(
                kernel=kernel,
                arguments=args,
                cancellation_token=cancellation_token
            )

            return result or []

        except Exception as e:
            logging.error(f"Detailed error: {str(e)}", exc_info=True)
            return []

        finally:
            # Clear recursion guard
            if kernel:
                kernel._in_completion = False

    async def get_streaming_chat_message_contents(
        self,
        chat_history: ChatHistory,
        execution_settings: Optional[PromptExecutionSettings] = None,
        kernel: Optional[Kernel] = None,
        cancellation_token=Optional[asyncio.Event]
    ) -> AsyncIterable[StreamingChatMessageContent]:
        raise NotImplementedError()
